File: router/dispatcher.py
"""Agent registry + DAG execution engine."""
from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import Awaitable, Callable, Optional

import httpx

from router.settlement import settle_agent_payment
from shared.config import ROUTER_WALLET
from shared.models import (
    AgentCapability,
    AgentInvocation,
    AgentRegistration,
    AgentRequest,
    PaymentRecord,
    Subtask,
)


class AgentRegistry:
    def __init__(self) -> None:
        self._agents: dict[str, AgentRegistration] = {}

    def register(self, agent: AgentRegistration) -> None:
        self._agents[agent.agent_id] = agent
        print(
            f"[registry] + {agent.name} ({agent.capability.value}) "
            f"@ ${agent.price_usd:.4f}/call -> {agent.endpoint}"
        )

    def unregister(self, agent_id: str) -> None:
        removed = self._agents.pop(agent_id, None)
        if removed:
            print(f"[registry] - {removed.name} ({agent_id})")

    def find_by_capability(
        self, capability: AgentCapability
    ) -> list[AgentRegistration]:
        return [a for a in self._agents.values() if a.capability == capability]

    def select_best(
        self, capability: AgentCapability
    ) -> Optional[AgentRegistration]:
        candidates = self.find_by_capability(capability)
        if not candidates:
            return None
        candidates.sort(key=lambda a: (a.price_usd, a.avg_latency_ms))
        return candidates[0]

    def list_all(self) -> list[AgentRegistration]:
        return list(self._agents.values())


registry = AgentRegistry()


async def invoke_agent(
    agent: AgentRegistration,
    subtask: Subtask,
    context: dict,
) -> tuple[AgentInvocation, str]:
    """POST an AgentRequest to the agent and return (invocation, output)."""
    invocation = AgentInvocation(
        subtask_id=subtask.subtask_id,
        agent_id=agent.agent_id,
        agent_name=agent.name,
        capability=agent.capability,
        input_data=subtask.input_data,
        price_usd=agent.price_usd,
        status="running",
        started_at=datetime.now(timezone.utc),
    )

    request = AgentRequest(
        invocation_id=invocation.invocation_id,
        input_data=subtask.input_data,
        context=context,
    )

    try:
        async with httpx.AsyncClient(timeout=30.0) as http:
            r = await http.post(
                f"{agent.endpoint.rstrip('/')}/invoke",
                json=request.model_dump(),
            )
            r.raise_for_status()
            payload = r.json()
            output = payload.get("output_data", "")
        invocation.output_data = output
        invocation.status = "completed"
    except Exception as e:
        output = f"[error] {type(e).__name__}: {e}"
        invocation.output_data = output
        invocation.status = "failed"
    finally:
        invocation.completed_at = datetime.now(timezone.utc)

    return invocation, output


def _topological_layers(subtasks: list[Subtask]) -> list[list[Subtask]]:
    """Group subtasks into parallel layers honoring `depends_on`."""
    by_id = {s.subtask_id: s for s in subtasks}
    remaining = set(by_id.keys())
    resolved: set[str] = set()
    layers: list[list[Subtask]] = []

    while remaining:
        layer_ids = [
            sid for sid in remaining
            if all(dep in resolved for dep in by_id[sid].depends_on)
        ]
        if not layer_ids:
            raise RuntimeError(
                f"Circular dependency detected in subtasks: {remaining}"
            )
        layer = [by_id[sid] for sid in layer_ids]
        layers.append(layer)
        resolved.update(layer_ids)
        remaining.difference_update(layer_ids)

    return layers


async def dispatch_subtasks(
    subtasks: list[Subtask],
    user_wallet: str,
    on_invocation_complete: Optional[
        Callable[[AgentInvocation, Optional[PaymentRecord]], Awaitable[None]]
    ] = None,
) -> tuple[list[AgentInvocation], list[PaymentRecord]]:
    """Execute a subtask DAG in topological layers, settling payments live."""
    all_invocations: list[AgentInvocation] = []
    all_payments: list[PaymentRecord] = []
    outputs_by_subtask: dict[str, str] = {}

    layers = _topological_layers(subtasks)

    for layer in layers:
        async def run_one(subtask: Subtask) -> None:
            agent = registry.select_best(subtask.required_capability)
            if agent is None:
                invocation = AgentInvocation(
                    subtask_id=subtask.subtask_id,
                    agent_id="none",
                    agent_name="<unavailable>",
                    capability=subtask.required_capability,
                    input_data=subtask.input_data,
                    price_usd=0.0,
                    status="failed",
                    output_data=(
                        f"No agent registered for capability "
                        f"{subtask.required_capability.value}"
                    ),
                    started_at=datetime.now(timezone.utc),
                    completed_at=datetime.now(timezone.utc),
                )
                all_invocations.append(invocation)
                outputs_by_subtask[subtask.subtask_id] = invocation.output_data or ""
                if on_invocation_complete:
                    await on_invocation_complete(invocation, None)
                return

            context = {
                dep_id: outputs_by_subtask.get(dep_id, "")
                for dep_id in subtask.depends_on
            }

            invocation, output = await invoke_agent(agent, subtask, context)
            outputs_by_subtask[subtask.subtask_id] = output

            payment: Optional[PaymentRecord] = None
            if invocation.status == "completed":
                payment = await settle_agent_payment(
                    from_wallet=ROUTER_WALLET,
                    to_wallet=agent.wallet_address,
                    amount_usd=agent.price_usd,
                    agent_id=agent.agent_id,
                    invocation_id=invocation.invocation_id,
                )
                invocation.payment_tx_hash = payment.tx_hash
                all_payments.append(payment)

            all_invocations.append(invocation)

            if on_invocation_complete:
                await on_invocation_complete(invocation, payment)

        await asyncio.gather(*(run_one(s) for s in layer))

    return all_invocations, all_payments

File: src/router/dispatcher.ts
// Agent registry + dependency-aware dispatcher.
// Every invocation calls the agent's /invoke via `fetchWithPayment`, which
// automatically satisfies the x402 402 response with a real USDC payment.

import {
  AgentCapability,
  AgentInvocation,
  AgentRegistration,
  AgentRequest,
  AgentResponse,
  PaymentRecord,
  Subtask,
} from "../shared/types.js";
import { config } from "../shared/config.js";
import { getFetchWithPayment, decodePaymentHeader, activeNetwork } from "./x402-client.js";

type InvocationCallback = (
  inv: AgentInvocation,
  payment: PaymentRecord | null,
) => void | Promise<void>;

// Optional injection point so tests can run the dispatcher against mocked
// fetch/signing without touching real x402 plumbing.
export interface DispatcherDeps {
  fetcher?: typeof globalThis.fetch;
  network?: string;
}

export class AgentRegistry {
  private agents = new Map<string, AgentRegistration>();

  register(agent: AgentRegistration): void {
    this.agents.set(agent.agentId, agent);
    console.log(
      `[registry] + ${agent.name} (${agent.capability}) @ $${agent.priceUsd.toFixed(4)}/call -> ${agent.endpoint}`,
    );
  }

  unregister(agentId: string): void {
    const removed = this.agents.get(agentId);
    this.agents.delete(agentId);
    if (removed) console.log(`[registry] - ${removed.name} (${agentId})`);
  }

  findByCapability(cap: AgentCapability): AgentRegistration[] {
    return [...this.agents.values()].filter((a) => a.capability === cap);
  }

  selectBest(cap: AgentCapability): AgentRegistration | undefined {
    const candidates = this.findByCapability(cap);
    candidates.sort((a, b) => a.priceUsd - b.priceUsd);
    return candidates[0];
  }

  listAll(): AgentRegistration[] {
    return [...this.agents.values()];
  }

  clear(): void {
    this.agents.clear();
  }
}

export const registry = new AgentRegistry();

export function topologicalLayers(subtasks: Subtask[]): Subtask[][] {
  const byId = new Map(subtasks.map((s) => [s.subtaskId, s]));
  const resolved = new Set<string>();
  const remaining = new Set(subtasks.map((s) => s.subtaskId));
  const layers: Subtask[][] = [];

  while (remaining.size > 0) {
    const layerIds = [...remaining].filter((id) =>
      (byId.get(id)!.dependsOn).every((d) => resolved.has(d)),
    );
    if (layerIds.length === 0) {
      throw new Error(
        `Circular dependency detected in subtasks: ${[...remaining].join(", ")}`,
      );
    }
    layers.push(layerIds.map((id) => byId.get(id)!));
    for (const id of layerIds) {
      resolved.add(id);
      remaining.delete(id);
    }
  }
  return layers;
}

export async function invokeAgent(
  agent: AgentRegistration,
  subtask: Subtask,
  context: Record<string, string>,
  deps: DispatcherDeps = {},
): Promise<{ invocation: AgentInvocation; output: string; payment: PaymentRecord | null }> {
  const invocation: AgentInvocation = {
    invocationId: crypto.randomUUID().slice(0, 8),
    subtaskId: subtask.subtaskId,
    agentId: agent.agentId,
    agentName: agent.name,
    capability: agent.capability,
    inputData: subtask.inputData,
    priceUsd: agent.priceUsd,
    status: "running",
    startedAt: new Date().toISOString(),
  };

  const fetcher = deps.fetcher ?? getFetchWithPayment();
  const net = deps.network ?? activeNetwork();

  const requestBody: AgentRequest = {
    invocationId: invocation.invocationId,
    inputData: subtask.inputData,
    context,
  };

  let payment: PaymentRecord | null = null;

  try {
    const res = await fetcher(`${agent.endpoint.replace(/\/$/, "")}/invoke`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(requestBody),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => "<no body>");
      throw new Error(`HTTP ${res.status}: ${text.slice(0, 200)}`);
    }

    const data = (await res.json()) as AgentResponse;
    invocation.outputData = data.outputData;
    invocation.status = "completed";
    invocation.completedAt = new Date().toISOString();

    // Pull the x402 payment receipt from response headers.
    const xPayResp = res.headers.get("X-PAYMENT-RESPONSE") ?? res.headers.get("x-payment-response");
    const decoded = decodePaymentHeader(xPayResp);
    if (decoded) {
      // The exact shape of the decoded object depends on x402 version; we
      // defensively read common fields.
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const d = decoded as any;
      invocation.x402PaymentResponse = d;
      invocation.paymentTxHash = d.transaction ?? d.txHash ?? d.tx_hash ?? "";
      payment = {
        txHash: invocation.paymentTxHash ?? "",
        fromWallet: config.routerWalletAddress,
        toWallet: agent.walletAddress,
        amountUsd: agent.priceUsd,
        agentId: agent.agentId,
        invocationId: invocation.invocationId,
        settledAt: new Date().toISOString(),
        chain: net,
        currency: "USDC",
        x402Scheme: "exact",
      };
      console.log(
        `[settle] $${agent.priceUsd.toFixed(4)} USDC -> ${agent.name} ` +
          `(${payment.txHash.slice(0, 14)}…) on ${net}`,
      );
    } else {
      // This should not happen once x402 is wired. If it does, surface it.
      console.warn(
        `[settle] WARNING: no X-PAYMENT-RESPONSE from ${agent.name}. ` +
          `Agent may not be x402-gated, or facilitator did not settle.`,
      );
    }

    return { invocation, output: invocation.outputData ?? "", payment };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    invocation.status = "failed";
    invocation.completedAt = new Date().toISOString();
    invocation.outputData = `[error] ${message}`;
    return { invocation, output: invocation.outputData, payment: null };
  }
}

export async function dispatchSubtasks(
  subtasks: Subtask[],
  onInvocationComplete?: InvocationCallback,
  deps: DispatcherDeps = {},
): Promise<{ invocations: AgentInvocation[]; payments: PaymentRecord[] }> {
  const invocations: AgentInvocation[] = [];
  const payments: PaymentRecord[] = [];
  const outputsBySubtask = new Map<string, string>();
  const layers = topologicalLayers(subtasks);

  for (const layer of layers) {
    const promises = layer.map(async (subtask) => {
      const agent = registry.selectBest(subtask.requiredCapability);
      if (!agent) {
        const inv: AgentInvocation = {
          invocationId: crypto.randomUUID().slice(0, 8),
          subtaskId: subtask.subtaskId,
          agentId: "none",
          agentName: "<unavailable>",
          capability: subtask.requiredCapability,
          inputData: subtask.inputData,
          priceUsd: 0,
          status: "failed",
          outputData: `No agent registered for capability ${subtask.requiredCapability}`,
          startedAt: new Date().toISOString(),
          completedAt: new Date().toISOString(),
        };
        invocations.push(inv);
        outputsBySubtask.set(subtask.subtaskId, inv.outputData ?? "");
        if (onInvocationComplete) await onInvocationComplete(inv, null);
        return;
      }

      const context: Record<string, string> = {};
      for (const dep of subtask.dependsOn) {
        context[dep] = outputsBySubtask.get(dep) ?? "";
      }

      const { invocation, output, payment } = await invokeAgent(agent, subtask, context, deps);
      invocations.push(invocation);
      outputsBySubtask.set(subtask.subtaskId, output);
      if (payment) payments.push(payment);
      if (onInvocationComplete) await onInvocationComplete(invocation, payment);
    });
    await Promise.all(promises);
  }

  return { invocations, payments };
}
